server.go

```diff
@@ -56,6 +56,7 @@ type mongoServer struct {
 	pingWindow         [6]time.Duration
 	info               *mongoServerInfo
 	maxSocketReuseTime time.Duration
+	activeSocketsCount int
```

Comment:
Isn't this equivalent to `len(server.liveSockets)`? This line seems to imply so: https://github.com/lyft/mgo/pull/6/files#diff-34c6b408d72845d076d47126c29948d1R255

If so, can we not track it separately and instead compute it?

Reply:
We're using activeSocketsCount as a placeholder to count `liveSockets` var.

Basically at [this](https://github.com/lyft/mgo/blob/5d0d4e79604660161de357e13cc649cb70c679be/server.go#L131) line we're assuming that you're under pool limit, but line above we released the lock one line above. Now at [this](https://github.com/lyft/mgo/blob/5d0d4e79604660161de357e13cc649cb70c679be/server.go#L142) line we're adding the newly created socket without checkin to see are still under pool limit or not.

In short, we're not doing double checking. Another way of doing this is more intrusive and that is to make use of socket states like for example just before releasing the [lock](https://github.com/lyft/mgo/blob/5d0d4e79604660161de357e13cc649cb70c679be/server.go#L130) you'd add placeholder socket with `state == connecting`. This way your pool limit would be up to date the moment you release the lock and you'd not need to double check (as you're not adding new socket to the list [here](https://github.com/lyft/mgo/blob/5d0d4e79604660161de357e13cc649cb70c679be/server.go#L142))

Reply:
My concern here is you are introducing a computed variable that must be kept updated with the true `len(server.liveSockets)`. I'd prefer and it's safer going forward to on solely rely on `len` instead of a separate field on server. This call is O(1) as the length information is stored on the slice value, so it's not like it's having to walk an array to determine its size or anything.

Reply:
Simply relying on len is not going to be enough. There is a right way but that is intrusive and would require more changes in mgo driver. I'm thinking of picking this up after halloween but for now this should work fine.